feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go

```diff
@@ -19,85 +19,117 @@ import (
 
-	// Group requests by PLMNID
-	// Use the longest possible PLMN ID (6), in the worst case we'll fragment requests more then needed, but the routing
-	// will still be logically correct
-	reqMap := map[string]*protos.UpdateSessionRequest{}
+	// Group requests by [PLMNID][GwServiceType]
+	// - PLMNID: Use the longest possible PLMN ID (6), in the worst case we'll fragment requests more then needed,
+	//   but the routing will still be logically correct
+	// - GwServiceType: Each request will be sent to either Session_Proxy or for N7_N40_proxy
+	reqMap := map[string]map[gateway_registry.GwServiceType]*protos.UpdateSessionRequest{}
 	for _, u := range r.GetUpdates() {
 		if u == nil {
 			continue
 		}
 		plmnid := getPlmnId6(u.GetCommonContext().GetSid().GetId())
-		req, ok := reqMap[plmnid]
-		if !ok || req == nil {
-			req = &protos.UpdateSessionRequest{Updates: []*protos.CreditUsageUpdate{u}}
-			reqMap[plmnid] = req
-		} else {
-			req.Updates = append(req.Updates, u)
+
+		// Group requests by PLMNID
+		pMap, ok := reqMap[plmnid]
+		if !ok || pMap == nil {
+			// Group requests by GwServiceType
+			addReqMapEntry(reqMap, plmnid)
 		}
+
+		// Append the Updates to the Session_Proxy or N7_N40_Proxy map request,
+		// based on the RAT Type
+		proxyType := getSessionControllerProxyType(u.GetCommonContext().GetRatType())
+		reqMap[plmnid][proxyType].Updates = append(reqMap[plmnid][proxyType].Updates, u)
 	}
 	for _, m := range r.GetUsageMonitors() {
 		if m == nil {
 			continue
 		}
 		plmnid := getPlmnId6(m.GetSid())
-		req, ok := reqMap[plmnid]
-		if !ok || req == nil {
-			req = &protos.UpdateSessionRequest{UsageMonitors: []*protos.UsageMonitoringUpdateRequest{m}}
-			reqMap[plmnid] = req
-		} else {
-			req.UsageMonitors = append(req.UsageMonitors, m)
+		pMap, ok := reqMap[plmnid]
+		if !ok || pMap == nil {
+			addReqMapEntry(reqMap, plmnid)
 		}
+
+		// Append the UsageMonitors to the Session_Proxy or N7_N40_Proxy map request,
+		// based on the RAT Type
+		proxyType := getSessionControllerProxyType(m.GetRatType())
+		reqMap[plmnid][proxyType].UsageMonitors = append(reqMap[plmnid][proxyType].UsageMonitors, m)
 	}
-	resultChan := make(chan *protos.UpdateSessionResponse, len(reqMap))
+
+	// Each request in reqMap is segregated into two requests and
+	// relays each request to respective proxies (session_proxy and n7n40_proxy).
+	// Hence length of resultChan is len(reqMap) * 2
+
+	resultChan := make(chan *protos.UpdateSessionResponse, len(reqMap)*2)
+	var numOfProxyRreq int
 
 	// send a separate Update request for each unique PLMN ID
-	for plmnid, req := range reqMap {
-		go func(plmnid string, req *protos.UpdateSessionRequest) {
-			client, ctx, cancel, err := s.getSessionProxyClient(ctx, plmnid)
-			if err != nil {
-				glog.Errorf("failed connect to Session Proxy for PLMNID '%s': %v", plmnid, err)
-				resultChan <- genUpdateErrorResp(req)
-				return
+	for plmnid, proxyMap := range reqMap {
+		for _, req := range proxyMap {
```

Comment:
the `_` i think it is the RAT type, you don't need to check it anymore

Reply:
@uri200 , here ```_``` represents gateway_registry.GwServiceType but not RATType
```
	FegSessionProxy gateway_registry.GwServiceType = "session_proxy"
	FegN7N40Proxy   gateway_registry.GwServiceType = "n7_n40_proxy"
```
Other than RATType_TGPP_NR, as all other RATType requests are forwarded to session_proxy (as referred below), I thought to segregate the requests based on GWServiceType rather than RATType.
```
func getSessionControllerProxyType(r protos.RATType) gateway_registry.GwServiceType {
	switch r {
	case protos.RATType_TGPP_NR:
		return FegN7N40Proxy
	default:
		return FegSessionProxy
	}
}
```
Please let me know your thoughts on this.

Reply:
@uri200 , please let me know if any more comments on this.

Reply:
```
reqMap := map[string]map[gateway_registry.GwServiceType]
```


In this loop you don't need RatType at all. You already used RatType to select  GwServiceType  at line 80

Function `getSessionControllerProxyClient` should never use RatType but `gwType`. That function should just only do the selection depending on `gwType` flag.

```
	for plmnid, proxyMap := range reqMap {
		for gwType, req := range proxyMap {
		...

		go func(plmnid string,  gwType gateway_registry.GwServiceType, eq *protos.UpdateSessionRequest) {
				var ratType protos.RATType

				client, ctx, cancel, err := s.getSessionControllerProxyClient(ctx, gwType, plmnid)
				if err != nil {
					glog.Errorf("failed connect to %s for PLMNID '%s': %v", gwType, plmnid, err)
					resultChan <- genUpdateErrorResp(req)
					return
				}
			....
		
```

The whole point of the multidimensional map is the dimensions is giving you the coordinates where to send the message

Reply:
@uri200 , Done. Please check